### PR TITLE
Build one test product for both Swift Testing and XCTest (with Swift Testing enabled by default.)

### PR DESCRIPTION
**Pull request by grynspan**
This PR refactors the previously-experimental Swift Testing support logic so that only a single build product is produced when using both XCTest and Swift Testing, and detection of Swift Testing usage is no longer needed at compile time.

On macOS, Xcode 16 is responsible for hosting Swift Testing content, so additional changes may be needed in Xcode to support this refactoring. Such changes are beyond the purview of the Swift open source project.

Resolves rdar://120864035.

### CODE REVIEW

File: Sources/swift-test/Entrypoint.swift [hunk 1]
```diff
@@ -11,10 +11,40 @@
...
 import Commands
+#if canImport(Darwin.C)
+private import Darwin.C
+#endif
 
 @main
 struct Entrypoint {
-    static func main() async {
+    static func main() async throws {
+#if canImport(Darwin.C)
```
Comment (grynspan) [inline Sources/swift-test/Entrypoint.swift:21]: This is _probably_ a temporary home for this hosting logic.

File: Sources/Build/LLBuildCommands.swift [hunk 2]
```diff
@@ -213,62 +213,76 @@ final class TestEntryPointCommand: CustomLLBuildCommand, TestBuildCommand {
...
+            \#(generateTestObservationCode(buildParameters: buildParameters))
+
+            import XCTest
+            \#(discoveryModuleNames.map { "import \($0)" }.joined(separator: "\n"))
+            #endif
+
+            @main
+            @available(*, deprecated, message: "Not actually deprecated. Marked as deprecated to allow inclusion of deprecated tests (which test deprecated functionality) without warnings")
+            struct Runner {
+                private static func testingLibrary() -> String {
```
Comment (grynspan) [inline Sources/Build/LLBuildCommands.swift:252]: There's probably a better way to disambiguate here. I am open to suggestions.

File: Sources/Commands/SwiftBuildCommand.swift [hunk 3]
```diff
@@ -175,23 +175,20 @@ public struct SwiftBuildCommand: AsyncSwiftCommand {
         }
 
         if case .allIncludingTests = subset {
-            func updateTestingParameters(of buildParameters: inout BuildParameters, library: BuildParameters.Testing.Library) {
+            func updateTestingParameters(of buildParameters: inout BuildParameters) {
```
Comment (grynspan) [inline Sources/Commands/SwiftBuildCommand.swift:178]: This helper function is probably redundant now.

File: Sources/Commands/SwiftTestCommand.swift [hunk 4]
```diff
@@ -339,11 +334,6 @@ public struct SwiftTestCommand: AsyncSwiftCommand {
                 .filteredTests(specifier: options.testCaseSpecifier)
                 .skippedTests(specifier: options.skippedTests(fileSystem: swiftCommandState.fileSystem))
 
-            // If there were no matches, emit a warning.
-            if tests.isEmpty {
-                swiftCommandState.observabilityScope.emit(.noMatchingTests)
```
Comment (grynspan) [inline Sources/Commands/SwiftTestCommand.swift:344]: This diagnostic is misleading if Swift Testing _does_ have matches. I propose punting on it for now until we have a good way to correlate results from both libraries.

**Comment by grynspan** [comment]
@swift-ci please test

File: Sources/Commands/SwiftTestCommand.swift [hunk 5]
```diff
@@ -856,20 +814,34 @@ final class TestRunner {
...
-            args = [xctestPath.pathString]
-            args += additionalArguments
-            args += [testPath.pathString]
-            return args
+            args += [xctestPath.pathString]
+        case .swiftTesting:
+            // FIXME: better way to get path to self
+            let toolPath = String(unsafeUninitializedCapacity: 2048) { buffer in
+                var count = UInt32(buffer.count)
+                _NSGetExecutablePath(buffer.baseAddress!, &count)
```
Comment (grynspan) [inline Sources/Commands/SwiftTestCommand.swift:828]: Is there a more idiomatic way to get the current process' path in SwiftPM I should use? This is idiomatic on Darwin, but pretty low-level (and I haven't attempted to handle arbitrarily long paths.)

Actually, since `swift-test` is probably _not_ the final home for the Swift Testing host executable (for one thing, it links to SwiftSyntax which causes runtime noise), it is probably moot?

File: Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift [hunk 6]
```diff
@@ -113,18 +109,14 @@ extension BuildParameters {
             }
         }
 
-        /// Which testing library to use for this build.
-        public var library: Library
```
Comment (grynspan) [inline Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift:117]: I should probably move the `Library` type out of `BuildParameters` (since it ain't one anymore) but I'm going to save that for a separate PR due to its potential impact to Xcode 16 betas.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
Mac CI failures are due to the `swift-test` command not recognizing its own new `--test-bundle-path` argument.

File: Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift (hunk 6 above)
Comment (grynspan) [reply Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift:117]: #7778

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
Note to self: https://github.com/swiftlang/swift/blob/a3bf1fc4c2ddc46f899b49a88898dee82d63f63e/lib/SIL/IR/SILDeclRef.cpp#L558C1-L559C1

**Comment by grynspan** [comment]
Working around https://github.com/swiftlang/swift/pull/69113 not being present in CI toolchains (?)

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Sources/Build/LLBuildCommands.swift [hunk 7]
```diff
@@ -213,62 +213,99 @@ final class TestEntryPointCommand: CustomLLBuildCommand, TestBuildCommand {
...
+                    // Fallback if not specified: run XCTest (legacy behavior)
+                    return "xctest"
+                }
 
-                import XCTest
-                \#(discoveryModuleNames.map { "import \($0)" }.joined(separator: "\n"))
+                #if os(Linux)
+                // FIXME: work around crash on Amazon Linux 2 when main function is async (rdar://128303921)
+                @_silgen_name("$ss13_runAsyncMainyyyyYaKcF")
+                private static func _runAsyncMain(_ asyncFun: @Sendable @escaping () async throws -> ())
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:285]: Very!

File: Sources/Build/LLBuildCommands.swift [hunk 8]
```diff
@@ -213,62 +213,99 @@ final class TestEntryPointCommand: CustomLLBuildCommand, TestBuildCommand {
...
+
+            #if \#(xctestImportCondition)
+            \#(generateTestObservationCode(buildParameters: buildParameters))
+
+            import XCTest
+            \#(discoveryModuleNames.map { "import \($0)" }.joined(separator: "\n"))
+            #endif
+
+            @main
+            @available(macOS 10.15.0, iOS 11.0, watchOS 4.0, tvOS 11.0, *)
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:268]: This code _can't_ run pre-10.15 in the first place, so I'm not sure what you're asking. Can you clarify?

File: Sources/Commands/SwiftTestCommand.swift [hunk 9]
```diff
@@ -871,11 +844,11 @@ final class TestRunner {
...
-    /// Executes and returns execution status. Prints test output on standard streams if requested
-    /// - Returns: Boolean indicating if test execution returned code 0, and the output stream result
-    func test(outputHandler: @escaping (String) -> Void) -> Bool {
-        (test(outputHandler: outputHandler) as Result) != .failure
-    }
+//    /// Executes and returns execution status. Prints test output on standard streams if requested
+//    /// - Returns: Boolean indicating if test execution returned code 0, and the output stream result
+//    func test(outputHandler: @escaping (String) -> Void) -> Bool {
+//        (test(outputHandler: outputHandler) as Result) != .failure
+//    }
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Commands/SwiftTestCommand.swift:851]: I suppose I can!

File: Sources/Build/LLBuildCommands.swift [hunk 10]
```diff
@@ -213,62 +213,99 @@ final class TestEntryPointCommand: CustomLLBuildCommand, TestBuildCommand {
...
+        // FIXME: work around crash on Amazon Linux 2 when main function is async (rdar://128303921)
+        let asyncMainKeyword = if context.productsBuildParameters.triple.isLinux() {
+          ""
+        } else {
+          "async"
+        }
+
+        stream.send(
+            #"""
+            #if \#(swiftTestingImportCondition)
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:256]: I was going for consistency with `xctestImportCondition` which is platform-specific.

File: Sources/Build/LLBuildCommands.swift [hunk 11]
```diff
@@ -213,62 +213,99 @@ final class TestEntryPointCommand: CustomLLBuildCommand, TestBuildCommand {
...
+            import Testing
+            #endif
+
+            #if \#(xctestImportCondition)
```
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:260]: I'm not sure I follow. This is not easier to read:

```swift
#if \#(switch buildParameters.testingParameters.testProductStyle {
case .entryPointExecutable:
  "canImport(XCTest)"
case .loadableBundle:
  "false"
})
if testingLibrary == "xctest" {
```

I could rename it to `canImportXCTest`, but it would be incorrectly named on Darwin.

[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:260]: I renamed it to `isXCTMainAvailable`, how's that then?

File: Sources/Build/LLBuildCommands.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:268]: > Why the `.0`?

Force of habit? I can remove the zeroes if they bother you.

**Comment by grynspan** [comment]
@swift-ci please test

File: Sources/Build/LLBuildCommands.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift]: Removed this one.

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Sources/Build/LLBuildCommands.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (grynspan) [reply Sources/Build/LLBuildCommands.swift:284]: I should elaborate.

The issue, as far as we know, occurs when the main function is `async` because it perturbs some state somewhere somehow. This function, historically, only invokes XCTest. The targets under test that are failing are not using Swift Testing (yet.) So this change preserves the synchronous main function signature _and_ ensures that if you're calling into XCTest, you never hop off the main thread/actor. At the same time, if you're calling into Swift Testing, the direct invocation of `_runAsyncMain()` allows the expected hop into async land to occur.

In other words, this makes the codepath for XCTest have the same concurrency semantics as it did before Swift Testing, which should keep the bug from occurring more than sporadically.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Review by dennisweissmann — APPROVED** [review]
I've looked through all the changes and didn't spot any obvious issues though I'm not at all familiar with this code so please take this approval with a grain of salt :)

**Comment by grynspan** [comment]
> I've looked through all the changes and didn't spot any obvious issues though I'm not at all familiar with this code so please take this approval with a grain of salt :)

Thanks Dennis—the more eyes the better given the size of the diff.

**Comment by grynspan** [comment]
Wait what

**Comment by grynspan** [comment]
Okay then.
